### PR TITLE
[Feature anywhere] Add annotation click handler

### LINKED ISSUE #3317: [Feature Anywhere] Support clicking & custom tooltip functionality from vega charts
Coming from #2880 

As part of the project design, we need to add custom functionality in the vega line charts such that:
- clicking on an annotation datapoint opens a 'View events' flyout
- rendering custom tooltips based on the type of datatpoint (regular or annotation) that is being hovered on
 
**Custom clicking:**
This is needed to open the View Events flyout when an annotation datapoint is clicked on. This can be done by adding an event listener in the Vega View to process click events within the chart canvas. Adding filtering on "event"/"annotation" datapoints (datapoints coming from `VisLayer`s) will need to be done, including logic for determining what is/isn't this type. Additionally, will need to be able to toggle clicking on/off, such that it is turned off in the view events flyout. Otherwise this would allow infinite opening of the flyout.

**Custom tooltips:**
Current mocks have custom data shown when hovering over a particular datapoint, including listing an aggregate view of the events that exist within the particular time bucket. Refer to the mocks in the meta issue for details: #2880. Note that custom tooltips will also require filtering on "event"/"annotation" datapoints.

Based on the current Vega View architecture in Dashboards, there is a single vega base view, which is where the tooltip handler and click handlers can be defined. These maybe can move into a new vega view that overrides functionality to include the custom functionality we want.

For an example on the entry points for enabling the custom functionality - see draft PR #3369

### PR DESCRIPTION
**Pull request by amsiglan**
### Description
This PR adds framework to register event handers with vega based visualizations by adding config and providing handler in the vis_augmenter plugin.

### Issues Resolved
#3317 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuarrrr** [comment]
@amsiglan There are some unit test failures to check and fix: https://github.com/opensearch-project/OpenSearch-Dashboards/actions/runs/4605088099/jobs/8136698336?pr=3777

**Review by lezzago — APPROVED** [review]

File: src/plugins/vis_augmenter/public/vega/constants.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export enum VisInteraction {
+  VIEW_EVENTS_FLYOUT = 'viewEventsFlyout',
+}
+
+export enum VisInteractionHandler {
+  HANDLE_POINT_IN_TIME_CLICK = 'visAugmenter.handlePointInTimeClick',
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/vega/constants.ts:11]: nit: suggest to include `EVENT` somewhere in the name so it helps differentiate it and prevent confusion of someone thinking this includes other points on the chart, like the regular generated points from the metrics, etc.

Something like `HANDLE_POINT_IN_TIME_EVENT_CLICK` or `HANDLE_EVENT_CLICK`?

File: src/plugins/vis_augmenter/public/vega/helpers.ts [hunk 2]
```diff
@@ -340,3 +345,43 @@ export const addPointInTimeEventsLayersToSpec = (
 
   return newSpec;
 };
+
+/**
+ * Interaction handling functions mapped by their action names.
+ */
+export const interactionHandlersByAction = {
+  [VisInteractionHandler.HANDLE_POINT_IN_TIME_CLICK]: (
+    _event: ScenegraphEvent,
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/vega/helpers.ts:354]: could you add some comments on what SceneGraphEvent means?

**Comment by ohltyler** [comment]
LGTM regarding the clicking event handling framework! As part of the issue #3317 , can you also add a handler for custom tooltips?

And as a follow-up, can we add a flag somewhere to disable a particular event (e.g., clicking) from happening via the vis config somewhere? As part of the view events flyout, we will want to disable clicking on the charts to prevent endless flyout opening.

File: src/plugins/vis_type_vega/public/data_model/types.ts [hunk 3]
```diff
@@ -115,6 +115,7 @@ export interface OpenSearchDashboards {
   type: string;
   renderer: Renderer;
   visibleVisLayers?: Map<VisLayerTypes, boolean>;
+  visInteractions?: Array<{ event: string; handlerName: string }>;
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/data_model/types.ts:118]: nit - can we name and export this type so we don't define it again in `vega_parser`?

File: src/plugins/vis_augmenter/public/vega/helpers.ts [hunk 4]
```diff
@@ -340,3 +345,43 @@ export const addPointInTimeEventsLayersToSpec = (
...
+
+  return {
+    ...config,
+    kibana: withInteractionsConfig,
+  };
+};
+
+export const isPointInTimeAnnotation = (item?: Item | null) => {
+  return item?.datum?.userAction === VisInteraction.VIEW_EVENTS_FLYOUT;
+};
```
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/vega/helpers.ts:387]: There's a naming mismatch here - the function name implies that it's checking if the item is a particular type (point in time annotation), whereas the actual check is whether the userAction is the view events flyout. Either we're checking the wrong thing or the function is named the wrong thing.

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js [hunk 5]
```diff
@@ -297,6 +298,17 @@ export class VegaBaseView {
         this._addDestroyHandler(() => tthandler.hideTooltip());
       }
 
+      if (this._parser.visInteractions) {
+        this._parser.visInteractions.forEach((interaction) => {
+          if (interactionHandlersByAction[interaction.handlerName]) {
+            view.addEventListener(
+              interaction.event,
+              interactionHandlersByAction[interaction.handlerName]
+            );
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/vega_view/vega_base_view.js:307]: nit - prefer destructuring

```suggestion
        this._parser.visInteractions.forEach((interaction) => {
          const { handlerName, event } = interaction;
          if (interactionHandlersByAction[handlerName]) {
            view.addEventListener(
              event,
              interactionHandlersByAction[handlerName]
            );
```

**Review by joshuarrrr — COMMENTED** [review]
This mostly looks good on first read-through - just some minor nits about naming, and I agree with @ohltyler's suggestions.

How can this be tested or verified, though? There don't really seem to be any test cases, but maybe this is something targeted for functional tests. In any case some screencasts would help for other reviewers.

File: src/plugins/vis_augmenter/public/vega/helpers.ts (hunk 2 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/helpers.ts:354]: This type comes from the Vega-typings. We can F12 into it to look at the definition of that type.

`export type ScenegraphEvent = MouseEvent | TouchEvent | KeyboardEvent;`

File: src/plugins/vis_augmenter/public/vega/constants.ts (hunk 1 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/constants.ts:11]: Changing to `VisInteractionEventHandlerName`

File: src/plugins/vis_augmenter/public/vega/helpers.ts (hunk 4 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/helpers.ts]: Updated logic to check for the annotation type

**Comment by amsiglan** [comment]
> LGTM regarding the clicking event handling framework! As part of the issue #3317 , can you also add a handler for custom tooltips? The content of the tooltips can be TODO for now and I can wrap up in a separate PR.
> 
> And as a follow-up, can we add a flag somewhere to disable a particular event (e.g., clicking) from happening via the vis config somewhere? As part of the view events flyout, we will want to disable clicking on the charts to prevent endless flyout opening.

For the disable piece, I think it would be more convenient to check whether the flyout is already open and then no-op inside the flyout code than trying to disable it via config. Thoughts?

**Comment by amsiglan** [comment]
> This mostly looks good on first read-through - just some minor nits about naming, and I agree with @ohltyler's suggestions.
> 
> How can this be tested or verified, though? There don't really seem to be any test cases, but maybe this is something targeted for functional tests. In any case some screencasts would help for other reviewers.

Part of the PR where spec is produced gets tested using existing jest tests while the piece where we handle the click and hover events will be better testing using functional tests in the PR where the flyout is shown

File: src/plugins/vis_augmenter/public/vega/constants.ts (hunk 1 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/vega/constants.ts:11]: The point here was to change the actual enum values to have `EVENT` in the naming, to help clarify that the clicking is happening on a point in time event, rather than an arbitrary datapoint within a point-in-time-enabled chart.

File: src/plugins/vis_augmenter/public/vega/helpers.ts (hunk 2 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/vega/helpers.ts:354]: Sure, I mostly meant adding a small comment for readability without having to do separate research / have set up in an IDE to find the definitions. Also, could you explain why we want this instead of simply `MouseEvent` (click)?

File: src/plugins/vis_augmenter/public/vega/helpers.ts [hunk 6]
```diff
@@ -350,14 +350,22 @@ export const addPointInTimeEventsLayersToSpec = (
...
-  [VisInteractionHandler.HANDLE_POINT_IN_TIME_CLICK]: (
+  [VisInteractionEventHandlerName.HANDLE_POINT_IN_TIME_CLICK]: (
     _event: ScenegraphEvent,
     item?: Item | null
   ) => {
     if (isPointInTimeAnnotation(item)) {
       // TODO: Show the events flyout
     }
   },
+  [VisInteractionEventHandlerName.HANDLE_POINT_IN_TIME_HOVER_IN]: (
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/vega/helpers.ts:361]: nit: same point here, if we can rename `HANDLE_POINT_IN_TIME_HOVER_IN` to include `EVENT` within the name, that can help clarify it.

**Comment by ohltyler** [comment]
> > LGTM regarding the clicking event handling framework! As part of the issue #3317 , can you also add a handler for custom tooltips? The content of the tooltips can be TODO for now and I can wrap up in a separate PR.
> > And as a follow-up, can we add a flag somewhere to disable a particular event (e.g., clicking) from happening via the vis config somewhere? As part of the view events flyout, we will want to disable clicking on the charts to prevent endless flyout opening.
> 
> For the disable piece, I think it would be more convenient to check whether the flyout is already open and then no-op inside the flyout code than trying to disable it via config. Thoughts?

My point with this is, we would need somewhere to pass logic to the chart here to disable clicks / no-op. Do you have a suggestion on how to do that based on your changes?

File: src/plugins/vis_augmenter/public/vega/helpers.ts (hunk 2 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/helpers.ts:354]: The standard event handler signature in vega-lite api is 
`export type EventListenerHandler = (event: ScenegraphEvent, item?: Item | null) => void;`. Hence, I used that instead of `MouseEvent`, etc.

Also, from general commenting perspective, they should be used to explain the reasoning, the WHY behind complicated logic, etc. and not something that the IDEs are meant to help out with. So, I am not sold out on the idea that we should explain what that type is, since that is the standard vega-lite type.

**Comment by ohltyler** [comment]
> > > LGTM regarding the clicking event handling framework! As part of the issue #3317 , can you also add a handler for custom tooltips? The content of the tooltips can be TODO for now and I can wrap up in a separate PR.
> > > And as a follow-up, can we add a flag somewhere to disable a particular event (e.g., clicking) from happening via the vis config somewhere? As part of the view events flyout, we will want to disable clicking on the charts to prevent endless flyout opening.
> > 
> > 
> > For the disable piece, I think it would be more convenient to check whether the flyout is already open and then no-op inside the flyout code than trying to disable it via config. Thoughts?
> 
> My point with this is, we would need somewhere to pass logic to the chart here to disable clicks / no-op. Do you have a suggestion on how to do that based on your changes?

Given some of the framework needed as part of the View Events flyout, we will be able to pass sufficient context to the chart here where we can disable the action.

File: src/plugins/vis_augmenter/public/vega/constants.ts (hunk 1 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/constants.ts]: Done

**Review by ohltyler — APPROVED** [review]
LGTM! Thanks again for helping out here!

**Comment by ashwin-pc** [comment]
I have a high level concern with this approach. Expressions today pass interaction data through `ExpressionRendererEvent`s. This bypasses that abstraction all together and calls for separate callbacks based on the spec. I think we should definitely provide a provision in the spec to define the events we want to support but rely on the Renderer Event bus to trigger and capture events using a common event bus. 

This is how all VisBuilder events for example are handled today irrespective of the underlying renderer:https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_builder/public/application/components/workspace.tsx#L119. It then takes this event data and retransmits it using UI actions that and UI action handler can then listen to and perform an action. 

The flow i see here that i would expect is something like this:
1. Use the existing vegalite selection spec to augment the desired selections we want to capture, or add functionality to extend the spec to help you define the props needed for the event. e.g. event name and captured data https://vega.github.io/vega-lite-v2/docs/selection.html
2. Add feature to vegalite expression function to pass these interactions to the Expression renderer event bus so that existing event handlers for the expression loader can pass these events to the UI Actions plugin and emmit them across the app
3. Register an UI actions handler for the alert actions in the Alerting plugin that will trigger a flyout on the Dashboard page with additional context for the event in question.

**Comment by ashwin-pc** [comment]
Also based on what i learnt from @joshuarrrr it looks like the vega expression function is already setup to handle the the click and brush events and my gut feeling is that its using a pipeline similar to the one ive described above. Doing a deep dive on how that works could probably answer how we want to implement the same here.

File: src/plugins/vis_augmenter/public/vega/helpers.test.ts [hunk 7]
```diff
@@ -32,6 +32,7 @@ import {
   TEST_RESULT_SPEC_MULTIPLE_VIS_LAYERS,
   TEST_RESULT_SPEC_SINGLE_VIS_LAYER,
   TEST_RESULT_SPEC_SINGLE_VIS_LAYER_EMPTY,
+  TEST_RESULT_SPEC_WITH_VIS_INTERACTION_CONFIG,
```
Comment (amsiglan) [inline src/plugins/vis_augmenter/public/vega/helpers.test.ts:35]: unwanted, removing in next iteration

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js [hunk 8]
```diff
@@ -297,6 +298,15 @@ export class VegaBaseView {
         this._addDestroyHandler(() => tthandler.hideTooltip());
       }
 
+      ['click', 'mouseover', 'mouseout'].forEach((eventName) => {
```
Comment (ohltyler) [inline src/plugins/vis_type_vega/public/vega_view/vega_base_view.js:301]: I see you have `click` and `mouseover` covered in the external action. For `mouseout`, should there be an added action to hide the tooltip I'm assuming?

File: src/plugins/vis_augmenter/public/actions/external_action_action.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    shouldAutoExecute: async () => true,
+    execute: async (context: ExternalActionActionContext) => {
+      if (isPointInTimeAnnotation(context.data.item)) {
+        if (context.data.event === 'click') {
+          // TODO: show events flyout
+        } else if (context.data.event === 'mouseover') {
+          // TODO: show custom tooltip
+        }
+      }
+    },
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/actions/external_action_action.ts:56]: Should we have these actions split into 2? One for triggering the flyout opening, and another for triggering the custom tooltip? Curious of thoughts from @ashwin-pc on this as well.

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js (hunk 8 above)
Comment (amsiglan) [reply src/plugins/vis_type_vega/public/vega_view/vega_base_view.js:301]: Yes, we will need it. Wanted to first get review about this way of doing things. Can add the handler pretty quick.

File: src/plugins/vis_augmenter/public/actions/external_action_action.ts (hunk 9 above)
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts:56]: Depends on whether we want to keep events from the visualization generic and have the handler distinguish based off of the data. This way we will keep the events minimal and avoid adding specific info into the general framework.

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts:56]: Yeah - I don't have a strong preference either way. Agreed we don't want to overload too specific of events here. Maybe one per `VisLayer` implementation makes sense - in this particular case, handling click/mouseover/mouseout actions for the `PointInTimeEventsVisLayer`.

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts:56]: On second thought, perhaps we do want to have these split up. We will need a mechanism for disabling the click event behavior, but still enabling the custom tooltip behavior in the view events flyout. Do you see a way we could pass that logic here? Or if we split into 2 events maybe (1 for custom click event, 1 for custom hover event), be able to disable one and enable the other?

**Review by ashwin-pc — COMMENTED** [review]
Based on the code change and some personal digging into this I think you might be able to make this wayy less specific to your implementation and a safer change. The way Kibana had done this seems to be using signals and vega expression functions. This is what i was able to figure out:

1. They register each of the [vega expression functions](https://vega.github.io/vega/docs/api/extensibility/#expressionFunction) in the base view file (But not in the class): https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/main/src/plugins/vis_type_vega/public/vega_view/vega_base_view.js#L56-L63
2. The registered functions when called by the [signals api](https://vega.github.io/vega/docs/signals/) in the spec forward the event to the instance specific handler defined in the base class [Ref](https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/main/src/plugins/vis_type_vega/public/vega_view/vega_base_view.js#L294-L313). This is usually done using the `on -> update` props for signals. I presume this is done so that the specific functions like brush have access to the correct `this` value.
3. The specific handler function then calls `vis` expression function that has access to the event bus and the `eventSubject` object that then passes the data to the event bus using the appropriate event name (e.g.applyFilter)

For your implementation you can keep it quite simple, something like this:

1. Add a new function [here](https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/6c11630fddc14542d1af84eb13ecdb28ca15c203/src/plugins/vis_type_vega/public/vega_view/vega_base_view.js#L49-L54) like `opensearchDashboardsTrigger: 'triggerEventHandler'` and define the event handler in the base class like 

```ts
async triggerEventHandler(triggerName, data) {
    this._triggerEvent({ name: triggerName, data });
  }
```

and also add a prop for the Vega Base view to pass `_triggerEvent` similar to `_applyFilter`

2. Then in [`vega_siualization.js`](https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/main/src/plugins/vis_type_vega/public/vega_visualization.js#L88) pass the `triggerEvent` from `this._vis.API.events.` the same way `applyFilter` does. 
3. Then in the [`vis.ts`](https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/main/src/plugins/visualizations/public/expressions/vis.ts#L80) file in expressions, add the `triggerEvent` function that can pass the name and data to the expression event bus (eventSubject)
4. From then on it should just be a matter of listening for the trigger in the AD plugin and showing the flyout based on it :) 

Now to use this new function you will have to use [signals](https://vega.github.io/vega/docs/signals/). You can refer to the vega functional test to see how these signals are written and used: https://github.com/ashwin-pc/OpenSearch-Dashboards/blob/main/test/functional/apps/visualize/_vega_chart.ts#L35-L48

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js (hunk 8 above)
Comment (amsiglan) [reply src/plugins/vis_type_vega/public/vega_view/vega_base_view.js]: Refactored logic to use signals instead which covers the click, mouseover and mouseout events

File: src/plugins/ui_actions/public/types.ts [hunk 10]
```diff
@@ -51,6 +52,10 @@ export interface VisualizeFieldContext {
   contextualFields?: string[];
 }
 
+export interface ExternalActionContext {
+  data: any;
```
Comment (ohltyler) [inline src/plugins/ui_actions/public/types.ts:56]: for clarity - can we define this more to be `data: { event: any, datum: any }`? Since this action encompasses both click and mouse events, I think it's ok to be this fine-grained. And it makes it more clear when defining an action, what the context is that is passed, since it will have to perform differently based on the `event`.

File: src/plugins/vis_augmenter/public/actions/external_action_action.ts (hunk 9 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts]: @amsiglan thanks. This makes sense to me in the context of the click event; I can defined a UIAction that listens for this trigger to open the view events flyout. But there's 2 things I'd like to clarify:
1. For the scenario of disabling the clicking, but still enabling the custom tooltips, how do you propose I do this? One solution I could see is altering `addVisEventSignalsToSpecConfig()` to only have `on` for the mouse events, and just omit the click event.
2. This is maybe a bit of a miss on my part when reviewing the latest revisions made, but for the tooltips, how should we get sufficient context passed such that a UIAction can render the tooltip correctly? The original proposal was making changes to the implemented [TooltipHandler](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/feature/feature-anywhere/src/plugins/vis_type_vega/public/vega_view/vega_tooltip.js#L58) class, which is using the out-of-the-box `tooltip()` function provided by vega to have a customized tooltip to render it OUI-style. It requires a lot of extra context, such as the container the vega view is in, the view itself, and other positioning parameters.

Comment (amsiglan) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts]: For 1. We don't need to disable the clicks, we could just check if the flyout is already open then just no-op in the event listener.
For 2. we will need to do a quick POC and check if the provided event has all the information we need. I see it has the view and the position related information
image

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts]: > We don't need to disable the clicks, we could just check if the flyout is already open then just no-op in the event listener.
How would you propose this? We don't persist global flyout state so how would we be able to make such a check within the UIAction?

Comment (ashwin-pc) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts]: @ohltyler Why do we need to disable click? what is the interaction pattern that we want to achieve with that?

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/actions/external_action_action.ts]: So user can't infinitely re-open the flyout when clicking - details in the issue https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3317

**Review by ohltyler — APPROVED** [review]
LGTM! We can keep the tracking issue open until tooltip handling is merged :)

**Review by ashwin-pc — COMMENTED** [review]
Nice this is definitely almost ready, just had a few questions

**Review by ashwin-pc — COMMENTED** [review]
Nice this is definitely almost ready, just had a few questions

File: src/plugins/vis_augmenter/public/vega/helpers.ts [hunk 11]
```diff
@@ -48,6 +50,30 @@ export const enableVisLayersInSpecConfig = (spec: object, visLayers: VisLayers):
...
 };
 
+/**
+ * Adds the signals which vega will use to trigger required events on the point in time annotation marks
+ */
+export const addVisEventSignalsToSpecConfig = (spec: object) => {
+  const config = get(spec, 'config', { kibana: {} });
+  const signals = {
+    ...(config.kibana.signals || {}),
+    [`${VisAnnotationType.POINT_IN_TIME_ANNOTATION}`]: [
```
[the remark this answers is not in the document]
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/vega/helpers.ts:60]: This is just a `signals` config that isn't directly tied to the spec. It's embedded within the existing `kibana` field in the vega lite spec that's processed to handle OSD-specific items in the vega parser.

From my understanding `signals` here is just a mapping of annotation type to a `signals` array (that will be added to the compiled `vega` spec later on)

File: src/plugins/vis_type_vega/public/data_model/vega_parser.ts [hunk 12]
```diff
@@ -323,6 +324,52 @@ The URL is an identifier only. OpenSearch Dashboards and your browser will never
         delete this.spec.autosize;
       }
     }
+
+    if (this._config?.signals) {
+      Object.entries(this._config?.signals).forEach(([markId, signals]: [string, any]) => {
```
[the remark this answers is not in the document]
Comment (ohltyler) [reply src/plugins/vis_type_vega/public/data_model/vega_parser.ts:329]: No. Since we control this `signals` value here, as it's set by us in `addVisEventSignalsToSpecConfig()`. `signals` here isn't the actual `signals` field per the vega spec, but just the config field name embedded within `kibana` config obj. See comment above.

File: src/plugins/vis_augmenter/public/vega/helpers.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (amsiglan) [reply src/plugins/vis_augmenter/public/vega/helpers.ts:60]: Yes, that is correct @ohltyler . In the vega parser we use the key "markId" to find the mark and then add then update the signal with that mark name.

**Review by ashwin-pc — APPROVED** [review]
Thanks for all the context and changes @amsiglan, this looks good to me!

File: src/plugins/visualizations/public/expressions/vis.ts [hunk 13]
```diff
@@ -99,6 +100,10 @@ export class ExprVis extends EventEmitter {
           if (!this.eventsSubject) return;
           this.eventsSubject.next({ name: 'applyFilter', data });
         },
+        externalAction: (data: any) => {
```
Comment (joshuarrrr) [inline src/plugins/visualizations/public/expressions/vis.ts:103]: Is this planned for functional testing, or can it be unit tested?

File: src/plugins/vis_type_vega/public/data_model/vega_parser.ts [hunk 14]
```diff
@@ -323,6 +324,52 @@ The URL is an identifier only. OpenSearch Dashboards and your browser will never
...
+        }
+      });
+    }
+  }
+
+  /**
+   * This method recursively looks for a mark that includes the given style.
+   * Returns undefined if it doesn't find it.
+   */
+  getMarkWithStyle(marks: any[], style: string): any {
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/data_model/vega_parser.ts:350]: Seems like some unit tests here would be nice to validate all the conditional branching.

**Review by joshuarrrr — APPROVED** [review]
We could use more testing of edge case handling.
